cpp/ycm/tests/CMakeLists.txt

```diff
@@ -77,13 +82,39 @@ add_executable( ${PROJECT_NAME}
                 ${SOURCES}
               )
 
+if ( MSVC )
+  # This is needed to compile tests with the gtest shared library
+  set_target_properties( ${PROJECT_NAME}
+                         PROPERTIES
+                         COMPILE_DEFINITIONS
+                         "GTEST_LINKED_AS_SHARED_LIBRARY=1" )
+
+  # Preprocessor definition used to export symbols for gmock tests
+  set_target_properties( ${SERVER_LIB} PROPERTIES DEFINE_SYMBOL "TESTS_EXPORTS" )
+
+  # Fix gtest build on MSVC 11. See:
+  # http://stackoverflow.com/questions/8274588/c2977-stdtuple-too-many-template-arguments-msvc11
+  if ( MSVC_VERSION EQUAL 1700 )
+    add_definitions( /D _VARIADIC_MAX=10 )
+  endif()
+
+  # Build gmock and ycm_core_tests targets in cmake ycm/tests folder
+  foreach( OUTPUTCONFIG ${CMAKE_CONFIGURATION_TYPES} )
+    string( TOUPPER ${OUTPUTCONFIG} OUTPUTCONFIG )
+    set_target_properties( ${GMOCK_LIBRARIES} PROPERTIES
+      RUNTIME_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${PROJECT_BINARY_DIR} )
+    set_target_properties( ${PROJECT_NAME} PROPERTIES
+      RUNTIME_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${PROJECT_BINARY_DIR} )
+  endforeach()
+endif()
+
 target_link_libraries( ${PROJECT_NAME}
+                       ${Boost_LIBRARIES}
```

Comment:
I am getting these errors without this:
```
BoostParts.lib(function.obj) : error LNK2005: "public: virtual unsigned int __cdecl boost::python::objects::py_function_impl_base::max_arity(void)const " (?max_arity@py_function_impl_base@objects@python@boost@@UEBAIXZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
BoostParts.lib(function.obj) : error LNK2005: "public: virtual __cdecl boost::python::objects::py_function_impl_base::~py_function_impl_base(void)" (??1py_function_impl_base@objects@python@boost@@UEAA@XZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
BoostParts.lib(errors.obj) : error LNK2005: "void __cdecl boost::python::throw_error_already_set(void)" (?throw_error_already_set@python@boost@@YAXXZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
BoostParts.lib(errors.obj) : error LNK2005: "public: virtual __cdecl boost::python::error_already_set::~error_already_set(void)" (??1error_already_set@python@boost@@UEAA@XZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
C:\Users\micbou\AppData\Local\Temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.exe : fatal error LNK1169: one or more multiply defined symbols found [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
```
[Here](https://msdn.microsoft.com/en-us/library/72zdcz6f.aspx)'s the documentation of the error. I am trying to understand why it happens.